### PR TITLE
Add sandbox method for task insertion to service LB and service disovery

### PR DESCRIPTION
**Pull request by sanimej**
Added a new sandbox API to add a task to the load balancer and service discovery. This can be used by the engine to make the task available only if the health check status is ok. It can also be used potentially by the swarm agent to gracefully pull a task from the backends and allow a grace period to drain the existing connections in the upgrade case.

This doesn't affect the driver plumbing. And the container can still discover other services in the network so that health check with dependency on other services can be supported.

Signed-off-by: Santhosh Manohar <santhosh@docker.com>

### CODE REVIEW

File: endpoint.go [hunk 1]
```diff
@@ -500,8 +500,10 @@ func (ep *endpoint) sbJoin(sb *sandbox, options ...EndpointOption) error {
 		return err
 	}
 
-	if e := ep.addToCluster(); e != nil {
-		log.Errorf("Could not update state for endpoint %s into cluster: %v", ep.Name(), e)
+	if sb.ingress {
```
Comment (mavenugo) [inline endpoint.go]: can you please explain why is this change needed ?
Infact, it looks like you are trying to add the service to the cluster only for ingress sandbox case ?

**Comment by mavenugo** [comment]
@sanimej just the 1 comment. rest of the changes LGTM

File: endpoint.go (hunk 1 above)
Comment (sanimej) [inline endpoint.go]: @mavenugo: for managed containers engine will call the Enable/Disable service API. Since the ingress sandbox is created specially its handled explicitly here. Alternative would be to call EnableService [here] (https://github.com/docker/docker/blob/master/daemon/network.go#L176)  after the `ingress-endpoint` join to the `ingress' sandbox.

Comment (mavenugo) [inline endpoint.go]: I think that is a better idea and makes the workflow very clear. 
@mrjana wdyt ?

Comment (mrjana) [inline endpoint.go]: I think we should make it the caller's responsibility to call `EnableService/DisableService` even for ingress sandbox. So we probably don't need to do any special case logic here. It's always up to the caller. In case of ingress the caller will call `EnableService` immediately after `Join` and call `DisableService` immediately before `Leave` call.

Comment (sanimej) [inline endpoint.go]: @mrjana Ok, I will remove specific `ingress` check here. But when the container is going down its probably ok to call only `Leave` without an explicit `DisableService`. In any case we have to be defensive against calls who doesn't explicitly call `DisableService`. That is why in `sbLeave` I didn't remove the `deleteFromCluster` call.

Comment (mrjana) [inline endpoint.go]: @sanimej Makes sense

**Comment by sanimej** [comment]
@mrjana @mavenugo: Instead of the sandbox once variable added an endpoint boolean and removed the special check for the ingress sandbox. PTAL

File: sandbox.go [hunk 2]
```diff
@@ -719,6 +725,34 @@ func (sb *sandbox) SetKey(basePath string) error {
 	return nil
 }
 
+func (sb *sandbox) EnableService() error {
+	for _, ep := range sb.getConnectedEndpoints() {
+		if !ep.serviceEnabled {
+			if e := ep.addToCluster(); e != nil {
+				return fmt.Errorf("Could not update state for endpoint %s into cluster: %v", ep.Name(), e)
```
Comment (dongluochen) [inline sandbox.go]: nit: `Could` -> `could`. I see lower case is used in the files.

File: sandbox.go [hunk 3]
```diff
@@ -719,6 +725,34 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+		}
+	}
+	return nil
+}
+
+func (sb *sandbox) DisableService() error {
+	for _, ep := range sb.getConnectedEndpoints() {
+		if ep.serviceEnabled {
+			if e := ep.deleteFromCluster(); e != nil {
+				return fmt.Errorf("Could not delete state for endpoint %s from cluster: %v", ep.Name(), e)
```
Comment (mrjana) [inline sandbox.go]: Same as above error. Error strings should start in lower case.

File: sandbox.go [hunk 4]
```diff
@@ -719,6 +725,34 @@ func (sb *sandbox) SetKey(basePath string) error {
 	return nil
 }
 
+func (sb *sandbox) EnableService() error {
+	for _, ep := range sb.getConnectedEndpoints() {
+		if !ep.serviceEnabled {
```
Comment (mrjana) [inline sandbox.go]: This should be lock protected and may be even set within the lock. We can unset if there is a failure.

File: sandbox.go [hunk 5]
```diff
@@ -719,6 +725,34 @@ func (sb *sandbox) SetKey(basePath string) error {
...
+			ep.serviceEnabled = true
+			ep.Unlock()
+		}
+	}
+	return nil
+}
+
+func (sb *sandbox) DisableService() error {
+	for _, ep := range sb.getConnectedEndpoints() {
+		if ep.serviceEnabled {
```
Comment (mrjana) [inline sandbox.go]: Needs to be protected by the lock.

**Comment by sanimej** [comment]
@mrjana Addressed the comments. PTAL

**Comment by mrjana** [comment]
LGTM

**Comment by mavenugo** [comment]
LGTM
